### DISCUSSION TITLE
Trouble Installing Private Repo

### DISCUSSION THREAD

**Discussion by bdsanfelippo** [community/community#26530] [Actions]
<p>How do I install a private repo in a GitHub Action? Currently, my GitHub Action fails on “npm ci”. Here is what I’ve tried:</p>
<p>I have a private repo within my organization.</p>
<p>I want to install that private repo from a GitHub Action that runs on an internal repo within my organization.</p>
<p>I created a Deployment Key on the private repo and added the public key.</p>
<p>On the public repo, I added a secret and attached the private key.</p>
<p>In my public repo, I added a workflow containing the following:</p>
<pre><code>steps:
            - name: Checkout
              uses: actions/checkout@v2
            - name: Use Node.js ${{ matrix.node-version }}
              uses: actions/setup-node@v1
              with:
                  node-version: ${{ matrix.node-version }}
                  registry-url: https://npm.pkg.github.com/
                  scope: "@my-org"
            - name: Add SSH Private Key for Private Repo Installation
              # See the following link for documentation on ssh-agent
              # https://github.com/marketplace/actions/webfactory-ssh-agent
              uses: webfactory/ssh-agent@v0.2.0
              with:
                  ssh-private-key: ${{ secrets.MY_PRIVATE_KEY}}
</code></pre>
<p>In my build output, I see that the key was successfully added. However, my “npm ci” always fails with unauthenticated for installing the private repo. </p>
<p>Has anyone encountered installing a private repo through github actions and webfactory/ssh-agent@v0.2.0? Is there any way to debug what is wrong or does anyone know if there is any other configuration I need?</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/bdsanfelippo">@bdsanfelippo</a> ,</p>
<p>If you’d intall a private repo, would you consider to checkout the private repo content for your step instead? Please check the usage of actions/checkout <a href="https://github.com/actions/checkout#checkout-multiple-repos-private" rel="nofollow noopener">here</a>.</p>
<p>Regarding the action ‘webfactory/ssh-agent@v0.2.0’, as <a href="https://github.com/webfactory/ssh-agent#why" rel="nofollow noopener">doc</a> mentioned: o<span>nce this has been set up, </span>git clone<span> commands using </span>ssh<span> URLs will </span><em>just work</em><span>. Also, running </span>ssh<span> commands to connect to other servers will be able to use the key.</span></p>
<p><span>Should use ssh command for your steps?</span></p>
<p><span>It’s recommened to share a sample repo with whole yaml for further investigation. </span></p>

**Comment by bdsanfelippo** [comment]
<p>Thanks <a class="mention" href="/weide-zhou">@weide-zhou</a> .</p>
<p>For the actions/checkout - it looks like it needs a PAT. I wanted to use a Deployment Key as that is specific to a single Repo whereas a PAT is user based and could be used more broadly.</p>
<p>I"m not sure I can force ssh for a “npm ci” call.  My entire yaml is:</p>
<pre><code>name: Validate

on:
    pull_request:
        types: [opened, reopened, edited, synchronize]

jobs:
    validate:
        runs-on: ubuntu-latest

        strategy:
            matrix:
                node-version: [12.x]

        steps:
            - name: Checkout
              uses: actions/checkout@v2
            - name: Use Node.js ${{ matrix.node-version }}
              uses: actions/setup-node@v1
              with:
                  node-version: ${{ matrix.node-version }}
                  registry-url: https://npm.pkg.github.com/
                  scope: "@my-org"
            - name: Add SSH Private Key for Private Repo Installation
              uses: webfactory/ssh-agent@v0.2.0
              with:
                  ssh-private-key: ${{ secrets.SSH_PRIVATE_KEY }}
            - run: npm ci
            - run: npm run lint
            - run: npm test
              env:
                  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
                  NPM_TOKEN: ${{ secrets.GITHUB_TOKEN }}
                  CI: true
</code></pre>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/bdsanfelippo">@bdsanfelippo</a> ,</p>
<p>Looks you didn’t get the private repo content after the action ‘webfactory/ssh-agent@v0.2.0’, before command ‘npm ci’. Please follow the action usage metioned <a href="https://github.com/webfactory/ssh-agent#why" rel="nofollow noopener">here</a>:</p>
<p><span>Once this has been set up, </span>git clone<span> commands using </span>ssh<span> URLs will </span><em>just work</em><span>. Also, running </span>ssh<span> commands to connect to other servers will be able to use the key.</span></p>
<pre><code>git clone git@github.com:YOUR_USERNAME/REPO_NAME.git
</code></pre>
<p>And you can use ‘ls -al’ to check the folder structure, ‘cd’ into the correct path for command ‘npm ci’ then.</p>
<p>Note: please also add persist-credentials: ‘false’ for actions/checkout.</p>

**Comment by bdsanfelippo** [comment, marked as answer]
<p>Solved it. In using webfactory/ssh-agent, it states that all ssh commands will just work. In order to force the npm install to use ssh, the dependency needs to be defined with a ssh link. For example:</p>
<pre><code>"dependencies": {
        "@myOrg/my-private-repo": "git+ssh://git@github.com/myOrg/my-private-repo.git"
    },
</code></pre>
<p><a class="mention" href="/mpdude">@mpdude</a>,</p>
<p>It would be helpful if this was documented in the webfactory/ssh-agent documentation.</p>
